python/src/udf.rs

```diff
@@ -1,32 +1,134 @@
 
-#[pyfunction]
-#[pyo3(signature = (udf, result_type, *args))]
-pub(super) fn call_udf<'py>(
-    py: Python<'py>,
-    udf: &'py PyAny,
-    result_type: &'py PyAny,
-    args: &'py PyTuple,
-) -> PyResult<&'py PyAny> {
-    let result_type = DataType::from_pyarrow(result_type)?;
-
-    // 1. Make sure we can convert each input to and from arrow arrays.
-    let mut udf_args = Vec::with_capacity(args.len() + 1);
-    udf_args.push(result_type.to_pyarrow(py)?);
-    for arg in args {
-        let array_data = ArrayData::from_pyarrow(arg)?;
-        let py_array: PyObject = array_data.to_pyarrow(py)?;
-        udf_args.push(py_array);
+        println!("Making python udf evaluator");
+        let args = info.args.iter().map(|arg| arg.value_ref.clone()).collect();
+
+        Box::new(PyUdfEvaluator {
+            args,
+            callable: self.callable.clone(),
+        })
     }
-    let args = PyTuple::new(py, udf_args);
-    let result = udf.call_method("run_pyarrow", args, None)?;
 
-    let array_data: ArrayData = ArrayData::from_pyarrow(result)?;
-    assert_eq!(array_data.data_type(), &result_type);
+    fn uuid(&self) -> &uuid::Uuid {
+        &self.uuid
+    }
+}
+
+impl Evaluator for PyUdfEvaluator {
+    fn evaluate(&mut self, info: &dyn RuntimeInfo) -> anyhow::Result<ArrayRef> {
+        println!("Evaluating Python UDF");
+        let args = self
+            .args
+            .iter()
+            .map(|value_ref| -> anyhow::Result<_> { Ok(info.value(value_ref)?.array_ref()?) })
+            .try_collect()?;
 
-    Ok(result)
+        let result = Python::with_gil(|py| -> anyhow::Result<ArrayRef> {
+            // todo:
+            println!("inside gil");
+            Ok(self.evaluate_py(py, args))
+        })?;
+        // FRAZ: handle python errors.
+        Ok(result)
+    }
+}
+
+impl PyUdfEvaluator {
+    fn evaluate_py<'py>(&self, py: Python<'py>, args: Vec<ArrayRef>) -> ArrayRef {
+        // self.callable.call1(py, (args,)).unwrap().extract(py).unwrap();
+        // let args: Vec<PyObject> = args
+        //     .iter()
+        //     .map::<anyhow::Result<_>, _>(|value_ref| {
+        //         let value: ColumnarValue = info.value(value_ref)?;
+        //         let array_ref: ArrayRef = value.array_ref()?;
+        //         let array_data: ArrayData = aray_ref.to_data();
+        //         let py_obj: PyObject = array_data.to_pyarrow(py)?;
+
+        //         Ok(py_obj)
+        //     })
+        //     .collect::<anyhow::Result<_>>()?;
+
+        // let locals = [("args", args)].into_py_dict(py);
+        // py.run(&self.code, None, Some(&locals))?;
+        // let result_py: PyObject = locals.get_item("ret").unwrap().into();
+
+        // let result_array_data: ArrayData = ArrayData::from_pyarrow(result_py.as_ref(py))?;
+
+        // // We can't control the returned type, but we can refuse to accept the "wrong" type.
+        // ensure!(
```

Comment:
We can also pass the expected type to Python and do some of the coercion in Python. That is sometimes beneficial -- consider we get a "numeric" Pandas array. If we say "convert to PyArrow" it will, but it may pick `f64`, `i64` or whatever type it wants. If we instead say "convert to a PyArrow array of type `i32`", it will try to do that.

This is why the UDF pattern had "calling" pass the expected type to Python, since it lets it try to satisfy the expectation. If we wait til here, the only thing we can do is fail if it isn't what we expect (or possibly cast).

Reply:
Cool -- I believe we are doing this in the `_converted_func` call, where we pass the `result_type` in now.